source/extensions/filters/http/oauth/filter.cc

```diff
@@ -0,0 +1,475 @@
+  std::time_t epoch = std::time(nullptr);
+  int expires;
+  try {
+    expires = std::stoi(expires_);
+  } catch (const std::exception&) {
+    return false;
+  }
+
+  return expires > epoch;
+}
+
+bool OAuth2CookieValidator::isValid() const { return hmacIsValid() && timestampIsValid(); }
+
+OAuth2Filter::OAuth2Filter(FilterConfigSharedPtr config,
+                           std::unique_ptr<OAuth2Client>&& oauth_client)
+    : validator_(std::make_shared<OAuth2CookieValidator>()), oauth_client_(std::move(oauth_client)),
+      config_(std::move(config)) {
+
+  oauth_client_->setCallbacks(*this);
+}
+
+const std::string& OAuth2Filter::bearerPrefix() const {
+  CONSTRUCT_ON_FIRST_USE(std::string, "bearer ");
+}
+
+std::string OAuth2Filter::extractAccessToken(const Http::RequestHeaderMap& headers,
+                                             const std::string& parameter_name) const {
+  ASSERT(headers.Path() != nullptr);
+
+  // Start by looking for a bearer token in the Authorization header.
+  const Http::HeaderEntry* authorization = headers.Authorization();
+  if (authorization != nullptr) {
+    const auto value = StringUtil::trim(authorization->value().getStringView());
+    const auto& bearer_prefix = bearerPrefix();
+    if (absl::StartsWithIgnoreCase(value, bearer_prefix)) {
+      const size_t start = bearer_prefix.length();
+      return std::string(StringUtil::ltrim(value.substr(start)));
+    }
+  }
+
+  // Check for the named query string parameter.
+  const auto path = headers.Path()->value().getStringView();
+  const auto params = Http::Utility::parseQueryString(path);
+  const auto param = params.find(parameter_name);
+  if (param != params.end()) {
+    return param->second;
+  }
+
+  return EMPTY_STRING;
+}
+
+/**
+ * primary cases:
+ * 1) user is signing out
+ * 2) /_oauth redirect
+ * 3) user is authorized
+ * 4) user is unauthorized
+ */
+Http::FilterHeadersStatus OAuth2Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+
+  // The following 2 headers are guaranteed for regular requests. The asserts are helpful when
+  // writing test code to not forget these important variables in mock requests
+  const Http::HeaderEntry* host_header = headers.Host();
+  ASSERT(host_header != nullptr);
+  host_ = host_header->value().getStringView();
+
+  const Http::HeaderEntry* path_header = headers.Path();
+  ASSERT(path_header != nullptr);
+  const absl::string_view path_str = path_header->value().getStringView();
+
+  // Keep track of how many initial 302s (to the auth server) we initially distribute.
+  // The goal is to keep this number in line with the number of successful logins at the conclusion
+  // of this filter.
+  const auto* user_agent_header = headers.UserAgent();
+  const absl::string_view user_agent_value =
+      user_agent_header != nullptr ? user_agent_header->value().getStringView() : EMPTY_STRING;
+  const std::vector<std::string> v = absl::StrSplit(user_agent_value, '.');
```

Comment:
i think using auto (or `std::vector<abls::string_view>`) here and doing `std::string(v[0])` below avoids creating strings for the whole vector, but i could be wrong here